File: 06-Abstract-Factory/src/go/factory.go
package main

// Abstract Factory — declares creation methods for each product in the family.
// NOTE: Go uses interfaces instead of abstract classes.
// Business logic is in a free function BuildUI() since Go interfaces can't have default methods.

type UIFactory interface {
	CreateButton() Button
	CreateCheckbox() Checkbox
	CreateMenu() Menu
}

// Concrete Factory — Windows family
type WindowsFactory struct{}

func (WindowsFactory) CreateButton() Button     { return WindowsButton{} }
func (WindowsFactory) CreateCheckbox() Checkbox { return WindowsCheckbox{} }
func (WindowsFactory) CreateMenu() Menu         { return WindowsMenu{} }

// Concrete Factory — Mac family
type MacFactory struct{}

func (MacFactory) CreateButton() Button     { return MacButton{} }
func (MacFactory) CreateCheckbox() Checkbox { return MacCheckbox{} }
func (MacFactory) CreateMenu() Menu         { return MacMenu{} }

File: 06-Abstract-Factory/src/rust/factory.rs
use crate::products::*;

// Abstract Factory trait — one method per product type
// NOTE: Rust returns trait objects (Box<dyn Trait>) since there's no inheritance
pub trait UIFactory {
    fn create_button(&self) -> Box<dyn Button>;
    fn create_checkbox(&self) -> Box<dyn Checkbox>;
    fn create_menu(&self) -> Box<dyn Menu>;
}

// Concrete Factory — Windows family
pub struct WindowsFactory;
impl UIFactory for WindowsFactory {
    fn create_button(&self) -> Box<dyn Button> { Box::new(WindowsButton) }
    fn create_checkbox(&self) -> Box<dyn Checkbox> { Box::new(WindowsCheckbox) }
    fn create_menu(&self) -> Box<dyn Menu> { Box::new(WindowsMenu) }
}

// Concrete Factory — Mac family
pub struct MacFactory;
impl UIFactory for MacFactory {
    fn create_button(&self) -> Box<dyn Button> { Box::new(MacButton) }
    fn create_checkbox(&self) -> Box<dyn Checkbox> { Box::new(MacCheckbox) }
    fn create_menu(&self) -> Box<dyn Menu> { Box::new(MacMenu) }
}
